chain/bitcoind_zqm_events.go

```diff
@@ -319,3 +402,189 @@ func (b *bitcoindZMQEvents) txEventHandler() {
 		}
 	}
 }
+
+// sequenceEventHandler reads sequnce events from the ZMQ block socket and
+// forwards them along to the current rescan clients.
+//
+// NOTE: This must be run as a goroutine.
+func (b *bitcoindZMQEvents) sequenceEventHandler() {
+	defer b.wg.Done()
+
+	log.Info("Started listening for bitcoind sequence notifications "+
+		"via ZMQ on", b.sequenceConn.RemoteAddr())
+
+	// Set up the buffers we expect our messages to consume. For sequence
+	// events, the body is structured as the following based on the type of
+	// message:
+	// <32-byte hash>C :                 Blockhash connected
+	// <32-byte hash>D :                 Blockhash disconnected
+	// <32-byte hash>R<8-byte LE uint> : Transactionhash removed from
+	// 	mempool for non-block inclusion reason
+	// <32-byte hash>A<8-byte LE uint> : Transactionhash added mempool
+	//
+	// We'll allocate a fixed data slice that we'll reuse when reading
+	// transactions from bitcoind through ZMQ. There's no need to recycle
+	// this slice (zero out) after using it, as further reads will
+	// overwrite the slice and we'll only be deserializing the bytes
+	// needed.
+	var (
+		eventHash [32]byte
+		eventType [1]byte
+	)
+
+	for {
+		// Before attempting to read from the ZMQ socket, we'll make
+		// sure to check if we've been requested to shut down.
+		select {
+		case <-b.quit:
+			return
+		default:
+		}
+
+		// Poll an event from the ZMQ socket.
+		var (
+			bufs = [][]byte{eventHash[:], eventType[:]}
+			err  error
+		)
+		bufs, err = b.sequenceConn.Receive(bufs)
+		if err != nil {
+			// EOF should only be returned if the connection was
+			// explicitly closed, so we can exit at this point.
+			if err == io.EOF {
+				return
+			}
+
+			// It's possible that the connection to the socket
+			// continuously times out, so we'll prevent logging
+			// this error to prevent spamming the logs.
+			netErr, ok := err.(net.Error)
+			if ok && netErr.Timeout() {
+				log.Trace("Re-establishing timed out ZMQ " +
+					"transaction connection")
+				continue
+			}
+
+			log.Errorf("Unable to receive ZMQ %v message: %v",
+				sequenceZMQCommand, err)
+			continue
+		}
+
+		// We have an event! We'll now ensure it is a sequence event,
+		// deserialize it, and report it to the different rescan
+		// clients.
+		eventCode := string(bufs[1])
+		switch eventCode {
+
+		// Block connected.
+		case `C`:
+
+		// Block disconnected.
+		case `D`:
+
+		// Transaction is removed from mempool but NOT because it's
+		// confirmed.
+		case `R`:
```

Comment:
I think both `R` and `A` need to read the mempool sequence int or else it will affect the next time we call `Receive`